ethereum/rpc/types/utils.go

```diff
@@ -215,11 +215,17 @@ func DecodeTx(clientCtx client.Context, txBz tmtypes.Tx) (sdk.Tx, uint64) {
 		return nil, 0
 	}
 
-	switch tx := tx.(type) {
-	case *evmtypes.MsgEthereumTx:
-		gasUsed = tx.GetGas() // NOTE: this doesn't include the gas refunded
-	case sdk.FeeTx:
-		gasUsed = tx.GetGas()
+	if len(tx.GetMsgs()) == 1 {
+		ethMsg, ok := tx.GetMsgs()[0].(*evmtypes.MsgEthereumTx)
+		if ok {
+			gasUsed = ethMsg.GetGas() // NOTE: this doesn't include the gas refunded
+			return tx, gasUsed
+		}
+	}
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if ok {
+		gasUsed = feeTx.GetGas()
 	}
```

Comment:
```suggestion
	feeTx, ok := tx.(sdk.FeeTx)
	if ok {
		return tx, feeTx.GetGas()
	}
	
	for _, msg := range tx.GetMsgs() {
		ethMsg, ok := msg.(*evmtypes.MsgEthereumTx)
		if ok {
			gasUsed = ethMsg.GetGas() // NOTE: this doesn't include the gas refunded
			return tx, gasUsed
		}
	}
```

Reply:
if we do that, we will always return 
`feeTx.GetGas()`

but in case of EthMsg, we want to ignore the tx fee to get the msg fee right?

Reply:
Nevermind, I completely refactor the logic because it wasnt clear. please check again